python/lib/core/dmod/core/dataset.py

```diff
@@ -487,6 +488,252 @@ def unlink_to_dataset(self, dataset: Dataset) -> bool:
         return dataset.manager is not None and dataset.manager.unlink_user(user=self, dataset=dataset)
 
 
+DataItem = TypeVar('DataItem', bound=Union[bytes, RepeatableReader, Path])
+
+
+class AbstractDomainDetector(ABC):
+    """ Abstraction for something that will automatically detect a :class:`DataDomain` for some data. """
+
+    def __init__(self, *args, **kwargs):
+        self._data_domain: Optional[DataDomain] = None
+
+    @abstractmethod
+    def detect(self, **kwargs) -> DataDomain:
+        """
+        Detect and return the data domain.
+
+        Parameters
+        ----------
+        kwargs
+            Optional kwargs applicable to the subtype, which may enhance or add to the domain detection and generation
+            capabilities, but which should not be required to produce a valid domain.
+
+        Returns
+        -------
+        DataDomain
+            The detected domain.
+
+        Raises
+        ------
+        DmodRuntimeError
+            If it was not possible to properly detect the domain.
+        """
+        pass
+
+    @property
+    def data_domain(self) -> DataDomain:
+        """
+        The domain detected by the instance, lazily initialized via :method:`detect` if needed.
+
+        Returns
+        -------
+        DataDomain
+            The domain detected by the instance.
+        """
+        if self._data_domain is None:
+            self._data_domain = self.detect()
+        return self._data_domain
+
+    def reset(self):
```

Comment:
Should all subtypes be required to implement this? Or can split this out someway?

Reply:
To implement `reset`?  No, if we are keeping the lazy property and it's attribute at the superclass level, then `reset` should implemented there also.  

If we aren't keeping the lazy property, then whether `reset` is appropriate or not depends on the details of each subclass.